pyparsing/diagram/__init__.py

```diff
@@ -294,40 +325,64 @@ def _to_diagram_element(
-    if isinstance(element, (pyparsing.Group, pyparsing.Forward)) and (
-        not element.customName or not exprs[0].customName
-    ):
-        # However, if this element has a useful custom name, we can pass it on to the child
-        if not exprs[0].customName:
-            propagated_name = name
-        else:
-            propagated_name = None
-
-        return _to_diagram_element(
-            element.expr,
-            parent=parent,
-            lookup=lookup,
-            vertical=vertical,
-            index=index,
-            name_hint=propagated_name,
-        )
+    if not element.customName:
+        if isinstance(element,
+                      (
+                          pyparsing.TokenConverter,
+                          pyparsing.Forward,
+                          pyparsing.Located,
+                      )
+                      ):
+            # However, if this element has a useful custom name, we can pass it on to the child
+            if not exprs[0].customName:
+                propagated_name = name
+            else:
+                propagated_name = None
+
+            return _to_diagram_element(
+                element.expr,
+                parent=parent,
+                lookup=lookup,
+                vertical=vertical,
+                index=index,
+                name_hint=propagated_name,
+                group_results_name=element_results_name,
+                show_results_names=show_results_names,
+            )
 
     # If the element isn't worth extracting, we always treat it as the first time we say it
     if _worth_extracting(element):
-        if el_id in lookup.first:
+        if el_id in lookup:
             # If we've seen this element exactly once before, we are only just now finding out that it's a duplicate,
             # so we have to extract it into a new diagram.
-            looked_up = lookup.first[el_id]
+            looked_up = lookup[el_id]
             looked_up.mark_for_extraction(el_id, lookup, name=name_hint)
-            return EditablePartial.from_call(railroad.NonTerminal, text=looked_up.name)
+            ret = EditablePartial.from_call(railroad.NonTerminal, text=looked_up.name)
+            if show_results_names and element_results_name:
+                ret = EditablePartial.from_call(
+                    railroad.Group, item=ret, label=element_results_name
+                )
+            return ret
 
         elif el_id in lookup.diagrams:
             # If we have seen the element at least twice before, and have already extracted it into a subdiagram, we
             # just put in a marker element that refers to the sub-diagram
-            return EditablePartial.from_call(
+            ret = EditablePartial.from_call(
                 railroad.NonTerminal, text=lookup.diagrams[el_id].kwargs["name"]
             )
+            if show_results_names and element_results_name:
+                ret = EditablePartial.from_call(
```

Comment:
I'm confused by this. Aren't we already creating a group here: https://github.com/pyparsing/pyparsing/blob/97fcd3dbf6947d9df445357a499c99d5ede7e5a1/pyparsing/diagram/__init__.py#L481-L484? 

I guess this second code chunk is now no longer being called. But actually I would rather this behaviour is put there instead of here, inside the extraction logic.

Reply:
I think there are two possible reasons why we might have a name annotation: if there is a defined results name on the current element, or if there was a name hint passed down from the parent. So we *might* see two annotations for the same element. I think this would do it:

    outer = Forward().setName("outer")
    inner = Word(alphas)[...].setName("inner")
    outer <<= inner

Rare, but not impossible.

Reply:
Right, but in that scenario the name hint should get overridden no? Hence it just being a "hint".

Reply:
Okay, I just commented out that second block, and generated the diagram for that snippet above, and it still generates the desired diagram. So I'll delete it.

Reply:
They are not quite exactly the same. One tests for the element_results_name and one tests for the group_results_name, potentially creating group annotations for both.

Reply:
I can see there's a snippet for that as well, but I believe these two are identical?

https://github.com/pyparsing/pyparsing/pull/298/files/97fcd3dbf6947d9df445357a499c99d5ede7e5a1#diff-bb7411dbf87d8b09728b60b573752816c42b0b4fc34f802d8209f75814a94b2bR381-R384

https://github.com/pyparsing/pyparsing/pull/298/files/97fcd3dbf6947d9df445357a499c99d5ede7e5a1#diff-bb7411dbf87d8b09728b60b573752816c42b0b4fc34f802d8209f75814a94b2bR481-R484

Reply:
They *are* identical. Code from line 481 has to be copied up to the earlier bits because of the early 'return' statements.

Reply:
Oh I see. Yes this redundancy bothers me a little. I wonder if that restructure wouldn't be too hard?

Reply:
I just looked at it again, there is at least one more more early return further down.

We *could* wrap it all in a try-finally, so that post return, this code always be done. But that is kind of an abuse of try-finally.

I'll try a decorator - it doesn't look so bad, and allows us to delete the duplicated code.

Reply:
Decorator looks good. I also just tried an experiment where we would show if a parse action is present - not an annotation, but another element in the sequence. I'm not adding it now, it was just to see how this would work, and this decorator is a nice place to hook in other types of such "embellishments".

Here's how it looks adding the parse action indicators (a block with a bold asterisk). It helps show why infixNotation creates the extra paths the way it does.
![Web capture_7-9-2021_104239_](https://user-images.githubusercontent.com/383875/132373609-4ddc3663-3778-4f67-8fe6-a81f9cbc9dee.jpeg)

Again, I'm not proposing this feature now. Just showing that we could use the decorator for other such items.

Reply:
I'll commit the decorator, see what you think.

Reply:
Ha! Once I moved the code to the decorator, the group_results_name turned out to be unused, so I just cut it. Less confusing cruft = better code.